File: 890-find-and-replace-pattern/890-find-and-replace-pattern.cpp
class Solution {
public:
    bool isSamePattern(string word, string pattern) {
        map<char, char> dic;
        map<char, char> inverse_dic;
        char c;
        for (int i=0; i<word.length(); i++) {
            c = word[i];
            if (dic.find(c) != dic.end()) {
                if (dic[c] != pattern[i]) {
                    return false;
                }
            }
            else {
                if (inverse_dic.find(pattern[i]) != inverse_dic.end()) {
                    if (inverse_dic[pattern[i]] != c) {
                        return false;
                    }
                }
                else {
                    dic[c] = pattern[i];
                    inverse_dic[pattern[i]] = c;
                }
            }
        }
        return true;
    }
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> res;
        for (auto const& w : words) {
            if (this->isSamePattern(w, pattern)) {
                res.push_back(w);
            }
        }
        return res;
    }
};

File: 890-find-and-replace-pattern/890-find-and-replace-pattern.py
class Solution:
    def isSamePattern(self, word: str, pattern: str) -> bool:
        dic = {}
        inverse_dic = {}
        for i, c in enumerate(word):
            if c in dic:
                if dic[c] != pattern[i]:
                    return False
            elif pattern[i] in inverse_dic:
                if inverse_dic[pattern[i]] != c:
                    return False
            else:
                dic[c] = pattern[i]
                inverse_dic[pattern[i]] = c
        return True
                
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        res = []
        for w in words:
            if self.isSamePattern(w, pattern):
                res.append(w)
        return res
